File: go/internal/service/product_service.go
package service

import (
	"errors"

	"github.com/app/gin-postgres-api/internal/models"
	"github.com/app/gin-postgres-api/internal/repository"
	"gorm.io/gorm"
)

type ProductService interface {
	GetAll() ([]models.Product, error)
	GetByID(id string) (*models.Product, error)
	Create(name, desc string, price int64, userID string) (*models.Product, error)
	Update(id string, name string, desc string, price int64) (*models.Product, error)
	Delete(id string) error
}

type ProductServiceImpl struct {
	repo repository.ProductRepository
}

func NewProductService(repo repository.ProductRepository) ProductService {
	return &ProductServiceImpl{repo: repo}
}

func (s *ProductServiceImpl) GetAll() ([]models.Product, error) {
	return s.repo.FindAll()
}

func (s *ProductServiceImpl) GetByID(id string) (*models.Product, error) {
	product, err := s.repo.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, errors.New("product not found")
	}
	return product, err
}

func (s *ProductServiceImpl) Create(name, desc string, price int64, userID string) (*models.Product, error) {
	product := &models.Product{
		Name:   name,
		Desc:   desc,
		Price:  price,
		UserID: userID,
	}
	if err := s.repo.Create(product); err != nil {
		return nil, err
	}
	return product, nil
}

func (s *ProductServiceImpl) Update(id string, name string, desc string, price int64) (*models.Product, error) {
	product, err := s.repo.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, errors.New("product not found")
	}
	if err != nil {
		return nil, err
	}

	if name != "" {
		product.Name = name
	}
	if desc != "" {
		product.Desc = desc
	}
	if price != 0 {
		product.Price = price
	}

	if err := s.repo.Update(product); err != nil {
		return nil, err
	}
	return product, nil
}

func (s *ProductServiceImpl) Delete(id string) error {
	_, err := s.repo.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return errors.New("product not found")
	}
	if err != nil {
		return err
	}
	return s.repo.Delete(id)
}

File: php/app/Services/ProductService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Repositories\ProductRepository;
use Ramsey\Uuid\Uuid;

class ProductService
{
    public function __construct(private ProductRepository $repo) {}

    public function getAll(): array
    {
        return $this->repo->findAll();
    }

    public function getById(string $id): ?array
    {
        return $this->repo->findById($id);
    }

    public function create(array $body, string $userId): array
    {
        $product = [
            'id'         => Uuid::uuid4()->toString(),
            'name'       => trim($body['name']),
            'desc'       => trim($body['desc'] ?? ''),
            'price'      => (int) ($body['price'] ?? 0),
            'user_id'    => $userId,
            'created_at' => (new \DateTime())->format(\DateTime::ATOM),
        ];

        return $this->repo->create($product);
    }

    public function update(string $id, array $body): bool
    {
        $existing = $this->repo->findById($id);
        if (!$existing) {
            return false;
        }

        $fields = array_filter([
            'name'  => isset($body['name']) ? trim($body['name']) : null,
            'desc'  => isset($body['desc']) ? trim($body['desc']) : null,
            'price' => isset($body['price']) ? (int) $body['price'] : null,
        ], fn($v) => $v !== null);

        if (empty($fields)) {
            return true;
        }

        return $this->repo->update($id, $fields);
    }

    public function delete(string $id): bool
    {
        return $this->repo->delete($id);
    }
}
